Sources/_StringProcessing/Algorithms/Matching/Matches.swift

```diff
@@ -183,13 +183,106 @@ extension BidirectionalCollection {
 
 // MARK: Regex algorithms
 
+@available(SwiftStdlib 5.7, *)
+struct RegexMatchesCollection<Output> {
+  let base: Substring
+  let regex: Regex<Output>
+  let startIndex: Index
+  
+  init(base: Substring, regex: Regex<Output>) {
+    self.base = base
+    self.regex = regex
+    self.startIndex = base.firstMatch(of: regex).map(Index.match) ?? .end
+  }
+}
+
+@available(SwiftStdlib 5.7, *)
+extension RegexMatchesCollection: Collection {
+  enum Index: Comparable {
+    case match(Regex<Output>.Match)
+    case end
+    
+    static func == (lhs: Self, rhs: Self) -> Bool {
+      switch (lhs, rhs) {
+      case (.match(let lhs), .match(let rhs)):
+        return lhs.range == rhs.range
+      case (.end, .end):
+        return true
+      case (.end, .match), (.match, .end):
+        return false
+      }
+    }
+    
+    static func < (lhs: Self, rhs: Self) -> Bool {
+      switch (lhs, rhs) {
+      case (.match(let lhs), .match(let rhs)):
+        // This implementation uses a tuple comparison so that an empty
+        // range `i..<i` will be ordered before a non-empty range at that
+        // same starting point `i..<j`. As of 2022-05-30, `Regex` does not
+        // return matches of this kind, but that is one behavior under
+        // discussion for regexes like /a*|b/ when matched against "b".
+        return (lhs.range.lowerBound, lhs.range.upperBound)
+          < (rhs.range.lowerBound, rhs.range.upperBound)
+      case (.match, .end):
+        return true
+      case (.end, .match), (.end, .end):
+        return false
+      }
+    }
+  }
+  
+  var endIndex: Index {
+    Index.end
+  }
+  
+  func index(after i: Index) -> Index {
+    let currentMatch: Element
+    switch i {
+    case .match(let match):
+      currentMatch = match
+    case .end:
+      fatalError("Can't advance past the 'endIndex' of a match collection.")
+    }
+    
+    let start: String.Index
+    if currentMatch.range.isEmpty {
+      if currentMatch.range.lowerBound == base.endIndex {
+        return .end
+      }
```

Comment:
Should this check be hoisted and/or asserted upon from non-empty ranges?

Reply:
If the current match's lower bound is the `endIndex`, what are some valid values for the upper bound?

Reply:
Oh, you're right - lowerBound == endIndex definitely signifies that we're at the end of matches.

Reply:
But even then it doesn't seem necessary to hoist that check - when the range is non empty the upper bound is always the departure point for the next search, and this condition isn't possible. This is just to prevent advancing past the end, and it goes away if we switch to the other "next match after empty" behavior

Reply:
IIUC, if you hoist it, it simplifies the code (it's an early exit) and you get the opportunity to add an internal consistency check, which is always very nice to have.